staging/src/k8s.io/apiserver/pkg/cel/library/format.go

```diff
@@ -0,0 +1,221 @@
+//
+// <Format>.validate(str: string) -> ?list<string>
+//
+//	Validates the given string against the given format. Returns optional.none
+//	if the string is valid, otherwise a list of validation error strings.
+//
+// <Format>.validatePrefix(str: string) -> ?list<string>
+//
+//	Same as validate, but assumes the input string will be used as a prefix for
+//	a kubernetes generated name.
+func Format() cel.EnvOption {
+	return cel.Lib(formatLib)
+}
+
+var formatLib = &format{}
+
+type format struct{}
+
+func (*format) LibraryName() string {
+	return "format"
+}
+
+func (*format) CompileOptions() []cel.EnvOption {
+	options := make([]cel.EnvOption, 0, len(formatLibraryDecls))
+	for name, overloads := range formatLibraryDecls {
+		options = append(options, cel.Function(name, overloads...))
+	}
+	for name, constantValue := range ConstantFormats {
+		options = append(options, cel.Constant("format."+name, apiservercel.FormatType, constantValue))
+	}
+	return options
+}
+
+func (*format) ProgramOptions() []cel.ProgramOption {
+	return []cel.ProgramOption{}
+}
+
+var ConstantFormats map[string]*apiservercel.Format = map[string]*apiservercel.Format{
+	"dns1123Label": {
+		Name:         "DNS1123Label",
+		ValidateFunc: func(s string) []string { return apimachineryvalidation.NameIsDNSLabel(s, false) },
+	},
+	"dns1123Subdomain": {
+		Name:         "DNS1123Subdomain",
+		ValidateFunc: func(s string) []string { return apimachineryvalidation.NameIsDNSSubdomain(s, false) },
+	},
+	"dns1035Label": {
+		Name:         "DNS1035Label",
+		ValidateFunc: func(s string) []string { return apimachineryvalidation.NameIsDNS1035Label(s, false) },
+	},
+	"qualifiedName": {
+		Name:         "QualifiedName",
+		ValidateFunc: validation.IsQualifiedName,
+	},
+
+	"dns1123LabelPrefix": {
+		Name:         "DNS1123LabelPrefix",
+		ValidateFunc: func(s string) []string { return apimachineryvalidation.NameIsDNSLabel(s, true) },
+	},
+	"dns1123SubdomainPrefix": {
+		Name:         "DNS1123SubdomainPrefix",
+		ValidateFunc: func(s string) []string { return apimachineryvalidation.NameIsDNSSubdomain(s, true) },
+	},
+	"dns1035LabelPrefix": {
+		Name:         "DNS1035LabelPrefix",
+		ValidateFunc: func(s string) []string { return apimachineryvalidation.NameIsDNS1035Label(s, true) },
+	},
+	"labelValue": {
+		Name:         "LabelValue",
+		ValidateFunc: validation.IsValidLabelValue,
+	},
+
+	// CRD formats
+	// Implementations sourced from strfmt, which kube-openapi uses as its
+	// format library. There are other CRD formats supported, but they are
+	// covered by other portions of the CEL library (like IP/CIDR), or their
+	// use is discouraged (like bsonobjectid, email, etc)
+	"uri": {
+		Name: "URI",
+		ValidateFunc: func(s string) []string {
+			// Directly call ParseRequestURI since we can get a better error message
+			_, err := url.ParseRequestURI(s)
+			if err != nil {
+				return []string{err.Error()}
+			}
+			return nil
+		},
+	},
```

Comment:
To confirm, we want to have both this and `isURL`?

Reply:
Thanks for pointing this out. The implementations are almost identical; there may be value in keeping this one since it returns the error string rather than a boolean. 

Looking thrugh native types for usages. I also did find that there were quite a few usages of `url.Parse` rather than `url.ParseRequestURI`. We currently have no way to express this specifically in CEL:

https://github.com/kubernetes/kubernetes/blob/6d4241fe5dc7341e3207b8ab96513ab1ddeae8a2/staging/src/k8s.io/apiserver/pkg/util/webhook/validation.go#L29-L32

https://github.com/kubernetes/kubernetes/blob/121607e80963370c1838f9f620c2b8552041abfc/staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go#L146-L157

https://github.com/kubernetes/kubernetes/blob/75695dae1093cc08cb56a4930c0be8e7e4433be1/staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation_encryption.go#L400-L409

Reply:
The ability to get an error has value. If we can document the relationship between `isURL` and the `url` format, that might be valuable since I can imagine someone using the format as a first line of defense and then needing to know if a unguarded `url(string)` call is safe so long as the format check passes.

Reply:
Added a note to the docs.